### PR TITLE
coprocessor: add and maintain bucket ndv for index histogram

### PR DESCRIPTION
**Pull request by winoros**
Signed-off-by: Yiding Cui <winoros@gmail.com>

### What problem does this PR solve?

to https://github.com/pingcap/tidb/pull/20580

Problem Summary:

When the analyze version goes to 2. We maintain the bucket ndv for each histogram bucket.

### What is changed and how it works?

- When appending new value to the bucket, the bucket's ndv += 1.
- When Merge the bucket. Merge the ndv between them.

### Related changes

- tidb pull request: https://github.com/pingcap/tidb/pull/20580

### Check List 

Tests 

- Unit test
- Integration test A good result on some large dataset whose size is > 50 million rows.

### Release note 
- No release note.

### CODE REVIEW

**Comment by zhongzc** [comment]
```
[2020-12-23T11:41:52.603Z] test coprocessor::test_analyze::test_analyze_index ... thread 'main' panicked at 'assertion failed: `(left == right)`
[2020-12-23T11:41:52.603Z]   left: `0`,
[2020-12-23T11:41:52.603Z]  right: `13`', tests/integrations/coprocessor/test_analyze.rs:232:5
[2020-12-23T11:41:52.603Z] stack backtrace:
[2020-12-23T11:41:52.603Z]    0: rust_begin_unwind
[2020-12-23T11:41:52.603Z]              at /rustc/b1496c6e606dd908dd651ac2cce89815e10d7fc5/library/std/src/panicking.rs:483:5
[2020-12-23T11:41:52.603Z]    1: std::panicking::begin_panic_fmt
[2020-12-23T11:41:52.603Z]              at /rustc/b1496c6e606dd908dd651ac2cce89815e10d7fc5/library/std/src/panicking.rs:437:5
[2020-12-23T11:41:52.603Z]    2: integrations::coprocessor::test_analyze::test_analyze_index
[2020-12-23T11:41:52.603Z]              at ./tests/integrations/coprocessor/test_analyze.rs:232:5
[2020-12-23T11:41:52.603Z]    3: integrations::coprocessor::test_analyze::test_analyze_index::{{closure}}
[2020-12-23T11:41:52.603Z]              at ./tests/integrations/coprocessor/test_analyze.rs:201:1
[2020-12-23T11:41:52.859Z]    4: core::ops::function::FnOnce::call_once
[2020-12-23T11:41:52.859Z]              at /rustc/b1496c6e606dd908dd651ac2cce89815e10d7fc5/library/core/src/ops/function.rs:227:5
[2020-12-23T11:41:52.859Z]    5: test_util::runner::run_test_with_hook::{{closure}}::{{closure}}
[2020-12-23T11:41:52.859Z]              at ./components/test_util/src/runner.rs:64:21
[2020-12-23T11:41:52.859Z]    6: core::ops::function::FnOnce::call_once{{vtable.shim}}
[2020-12-23T11:41:52.859Z]              at /rustc/b1496c6e606dd908dd651ac2cce89815e10d7fc5/library/core/src/ops/function.rs:227:5
[2020-12-23T11:41:52.859Z]    7: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
[2020-12-23T11:41:52.859Z]              at /rustc/b1496c6e606dd908dd651ac2cce89815e10d7fc5/library/alloc/src/boxed.rs:1042:9
[2020-12-23T11:41:52.859Z] note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[2020-12-23T11:41:52.859Z] FAILED
```

CI failed.

File: src/coprocessor/statistics/histogram.rs [hunk 1]
```diff
@@ -111,6 +140,36 @@ impl Histogram {
...
+
+        // create a new bucket and insert data
+        let mut count = 1;
+        if let Some(bucket) = self.buckets.last() {
+            count += bucket.count
+        }
+        self.buckets
+            .push(Bucket::new_v2(count, data.to_vec(), data.to_vec(), 1, 1));
+    }
+
```
Comment (zhongzc) [inline src/coprocessor/statistics/histogram.rs:172]: Seems a 99% copy of `append()`.

File: src/coprocessor/statistics/histogram.rs [hunk 2]
```diff
@@ -21,6 +23,23 @@ impl Bucket {
...
+        lower_bound: Vec<u8>,
+        repeats: u64,
+        ndv: u64,
+    ) -> Bucket {
+        Bucket {
+            count,
+            upper_bound,
+            lower_bound,
+            repeats,
+            ndv,
```
Comment (andylokandy) [inline src/coprocessor/statistics/histogram.rs:42]: `new_v2` is not necessary. You just need to add `ndv: u64` to `new()`

File: src/coprocessor/statistics/histogram.rs [hunk 3]
```diff
@@ -37,12 +56,22 @@ impl Bucket {
...
 
+    // insert a item bigger than current upper_bound,
+    // we need data to set as upper_bound
+    fn append_v2(&mut self, data: Vec<u8>) {
+        self.upper_bound = data;
+        self.count += 1;
+        self.repeats = 1;
+        self.ndv += 1;
+    }
+
```
Comment (andylokandy) [inline src/coprocessor/statistics/histogram.rs:67]: ditto. it's not necessarily duplicated.

**Review by zhongzc — APPROVED** [review]

**Review by andylokandy — APPROVED** [review]
LGTM

**Pull request merged**